lahja/trio/endpoint.py

```diff
@@ -0,0 +1,905 @@
+                buffer.extend(data)
+
+            msg = cast(Message, pickle.loads(buffer[4:t_size]))
+            await self._msg_send_channel.send(msg)
+            buffer = buffer[t_size:]
+
+    async def _cleanup(self) -> None:
+        await self._msg_send_channel.aclose()
+        await self._socket.aclose()
+
+    #
+    # Connection API
+    #
+    @classmethod
+    async def connect_to(cls, path: pathlib.Path) -> "TrioConnection":
+        socket = await trio.open_unix_socket(str(path))
+        cls.logger.debug("Opened connection to %s: %s", path, socket)
+        return cls(socket)
+
+    async def send_message(self, message: Msg) -> None:
+        msg_data = pickle.dumps(message)
+        size = len(msg_data)
+        async with self._lock:
+            try:
+                # TODO: look into buffering the writes
+                await self._socket.send_all(size.to_bytes(4, "little") + msg_data)
+            except (trio.ClosedResourceError, trio.BrokenResourceError) as err:
+                raise RemoteDisconnected from err
+
+    async def read_message(self) -> Message:
+        try:
+            return await self._msg_receive_channel.receive()
+        except trio.EndOfChannel as err:
+            raise RemoteDisconnected() from err
+
+
+class TrioRemoteEndpoint(BaseRemoteEndpoint):
+    def __init__(
+        self,
+        local_name: str,
+        conn: ConnectionAPI,
+        subscriptions_changed: ConditionAPI,
+        new_msg_func: Callable[[Broadcast], Awaitable[Any]],
+    ) -> None:
+        super().__init__(local_name, conn, new_msg_func)
+
+        self._notify_lock = trio.Lock()  # type: ignore
+
+        self._received_response = trio.Condition()  # type: ignore
+        self._received_subscription = trio.Condition()  # type: ignore
+
+        self._running = trio.Event()  # type: ignore
+        self._stopped = trio.Event()  # type: ignore
+
+        self._received_subscription = subscriptions_changed
+
+        self._subscriptions_initialized = trio.Event()  # type: ignore
+
+        self._running = trio.Event()  # type: ignore
+        self._stopped = trio.Event()  # type: ignore
+        self._ready = trio.Event()  # type: ignore
+
+    @asynccontextmanager
+    async def run(self) -> AsyncIterator[RemoteEndpointAPI]:
+        async with trio.open_nursery() as nursery:
+            await self._start(nursery)
+            try:
+                yield self
+            finally:
+                await self.stop()
+
+    async def _run(self) -> None:
+        async with cast(TrioConnection, self.conn).run():
+            await self._process_incoming_messages()
+
+    async def _start(self, nursery: trio_typing.Nursery) -> None:
+        if self.is_running:
+            raise LifecycleError("RemoteEndpoint is already running")
+        elif self.is_stopped:
+            raise LifecycleError("RemoteEndpoint has already been run and stopped")
+
+        nursery.start_soon(self._run)
```

Comment:
It looks like `trio.Nursery.start` is meant to designed to make this pattern easier to write: https://trio.readthedocs.io/en/latest/reference-core.html#trio.Nursery.start

Reply:
Yeah, remember that from the docs.  I'll open an issue to track refactoring this to use something like that pattern.